docs/OnTrack/Jupyter Notebook/Docker-Documentation-Research-T1-2022.md

```diff
@@ -1,32 +1,67 @@
+file, it will 'spin up' the Jupyter to PDF conversion container to perform the conversion process.
+
+This is done via a shell command that does several things:
+
+1. Firstly, it ensures that the file to be converted is renamed to "input.ipynb".
+2. It then instructs the container to run AND it mounts the file's directory as a Docker volume. As
+   the container runs it will execute the conversion process via its `ENTRYPOINT` command. The
+   converted file will be outputted to the volume. After the container has run, it will be removed.
+3. Finally, it removes the temporary files to ensure there is always free space.
+
+The `docker run` command allows us several options for this type of use-case: we are able to specify
+a volume using the `-v` option, and we are able to ask that the container is removed after it has
+finished its process with the `--rm` option.
 
-The API container will then: tell the conversion container to start, perform the conversion process,
-and then `exit`. Note that the conversion process will read from the _temp_ directory and write the
-output to the _temp_ directory.
+**Note: we may need to have some process in place to read whether the conversion was a success or a
+failure.**
 
-The entire bash script might look something like this:
+## Conversion Container Requirements
 
-```shell
-cp input-file /temp-directory # copy the input file into the shared directory
-docker run --rm conversion-image bash -c "convert temp-directory/input-file" # run the conversion container with the command to convert
-rm temp-directory/input-file # remove the input file after it has been converted
-```
+For a container to work in this architecture, there are some requirements that have to be met:
 
-**How do we tell the API container whether it was a success or failure??? Need more research into
-this**
+- The container will run the conversion as its `ENTRYPOINT` command _without any user input
+  required_.
+- The container will always look in a single directory (mounted as a volume when the container is
+  run) for **one** file that will be called "input" (+ whatever file extension is required).
+- The container will always output the converted file to the **same** directory (mounted as a volume
+  when the container is run) and call it "output.pdf".
 
-If the file conversion was a success, the API container will then read the output file from the temp
-directory and pass it on to the next steps. The files in the _temp_ directory can now be safely
-removed by the API container to free up space for the next conversion.
+These requirements allow for the containers themselves to remain isolated, while the logic about
+which files need to be converted can be handled seperately by OnTrack's backend. This is necessary
+to follow the Single-responsibility Principle: the container itself is responsible for only one
+task - that is, performing the conversion process.
```

Comment:
Comments for docker_flow.png:
- It may be worth changing 'frontend' and 'backend' to 'Front-end' and 'Back-end' respectively, to ensure readability and keep the same spelling throughout.
- Not sure what **POST** means on POST input file (arrow left of OnTrack Backend box). Not sure if this needs to be explained or not. 
- Whilst I understand what "return success or failure" means in that it says whether the conversion was a success or not, other viewers may not understand. It might be worth saying something like "return success or failure of conversion process", just so readers know what it indicates.

Just suggestions though.